filebeat/input/filestream/internal/input-logfile/harvester_test.go

```diff
@@ -0,0 +1,286 @@
+
+		for goroutinesBeforeRun < runtime.NumGoroutine() {
+			time.Sleep(1 * time.Millisecond)
+		}
+		assert.Nil(t, hg.StopGroup())
+	})
+
+	t.Run("assert a harvester error is handled", func(t *testing.T) {
+		mockHarvester := &mockHarvester{0, true}
+		hg := testDefaultHarvesterGroup(t, mockHarvester)
+
+		goroutinesBeforeRun := runtime.NumGoroutine()
+
+		hg.Start(input.Context{Logger: logp.L(), Cancelation: context.Background()}, source)
+
+		for goroutinesBeforeRun < runtime.NumGoroutine() {
+			time.Sleep(1 * time.Millisecond)
+		}
+
+		_, ok := hg.readers.table[source.Name()]
+		assert.False(t, ok)
+
+		err := hg.StopGroup()
+		assert.Error(t, err)
+	})
+
+	t.Run("assert already locked resource has to wait", func(t *testing.T) {
+		mockHarvester := &mockHarvester{}
+		hg := testDefaultHarvesterGroup(t, mockHarvester)
+		inputCtx := input.Context{Logger: logp.L(), Cancelation: context.Background()}
+
+		goroutinesBeforeRun := runtime.NumGoroutine()
+		r, err := lock(inputCtx, hg.store, source.Name())
+		if err != nil {
+			t.Fatalf("cannot lock source")
+		}
+
+		hg.Start(inputCtx, source)
+
+		locked := true
+		for goroutinesBeforeRun < runtime.NumGoroutine() {
+			time.Sleep(10 * time.Millisecond)
+
+			if locked {
+				releaseResource(r)
+				locked = false
+			}
+		}
+
+		assert.Equal(t, 1, mockHarvester.runCount)
+		assert.Nil(t, hg.StopGroup())
+	})
+
+	t.Run("assert already locked resource has no problem when harvestergroup is cancelled", func(t *testing.T) {
+		mockHarvester := &mockHarvester{}
+		hg := testDefaultHarvesterGroup(t, mockHarvester)
+		inputCtx := input.Context{Logger: logp.L(), Cancelation: context.Background()}
+
+		goroutinesBeforeRun := runtime.NumGoroutine()
+		r, err := lock(inputCtx, hg.store, source.Name())
+		if err != nil {
+			t.Fatalf("cannot lock source")
+		}
+		defer releaseResource(r)
+
+		hg.Start(inputCtx, source)
+
+		for goroutinesBeforeRun < runtime.NumGoroutine() {
+			time.Sleep(100 * time.Millisecond)
+
+			err := hg.StopGroup()
+			assert.Error(t, err)
+		}
+
+		assert.Equal(t, 0, mockHarvester.runCount)
+	})
+}
+
+func testDefaultHarvesterGroup(t *testing.T, mockHarvester Harvester) *defaultHarvesterGroup {
+	return &defaultHarvesterGroup{
+		readers:   newReaderGroup(),
+		pipeline:  &pipelinemock.MockPipelineConnector{},
+		harvester: mockHarvester,
+		store:     testOpenStore(t, "test", nil),
+		tg:        unison.TaskGroup{},
+	}
+
+}
+
+type mockHarvester struct {
+	runCount  int
+	returnErr bool
+}
+
+func (m *mockHarvester) Run(ctx input.Context, s Source, c Cursor, p Publisher) error {
+	m.runCount += 1
+
+	timed.Wait(ctx.Cancelation, 1*time.Second)
+
+	if m.returnErr {
+		return fmt.Errorf("harvester error")
+	}
+	return nil
+}
+
+func (m *mockHarvester) Test(_ Source, _ input.TestContext) error { return nil }
+
+func (m *mockHarvester) Name() string { return "mock" }
+
+type mockPanicHarvester struct{}
```

Comment:
Introducing a callback into the mockHarvester means that we do not need this type.